share/wake/lib/core/list.wake

```diff
@@ -889,3 +889,62 @@ export def groupBy (cmpFn: a => a => Order) (list: List a): List (List a) =
         _ -> (elem, Nil), acc
 
     foldr combo Nil sorted
+
+# Returns a list of items where each item was a memeber of the two parameter lists.
+# The returned list will not be the same order as the parameters.
+#
+# Guarantees:
+#  - Output list length will be <= the length of the smaller input list
+#  - All items in output list will be present in both input lists
+#  - Performance is considered and exceeds the naive N^2 algorithm
+#
+# Examples:
+# ```
+#   intersect icmp (seq 3) (seq 4) = (0, 1, 2, Nil)
+#   intersect icmp (seq 3) (seq 3) = (0, 1, 2, Nil)
+#   intersect icmp (seq 3) (4, 2, 0, Nil) = (0, 2, Nil)
+#   intersect scmp ("", "bar", "bat", Nil) ("", "foo", "bat", Nil) = ("", "bat", Nil)
+# ```
+export def intersect (cmpFn: a => a => Order) (lhs: List a) (rhs: List a): List a =
+    def lhsSort = sortBy cmpFn lhs
+    def rhsSort = sortBy cmpFn rhs
```

Comment:
You probably weighed this already, but we could also maintain the ordering according to one of the lists by turning the other into a Tree and testing membership; at that point we've got `O((n+m) log n)` or thereabouts, which is probably not too bad of a complexity, either.  Though if we still want to limit repeated members (and I do think we do) with that implementation, we'd probably need to go through Map with the value being a counter, which would be a bit trickier than the very straightforward head-comparison you're able to use here.

Reply:
I think its worth keeping the simpler implementation here. Its not typical for intersect to maintain order and I don't know of any code I've seen that needs both order and intersect.

The other question is _which_ order is correct. Both lists could have completely different orders so you have to arbitrarily pick on to maintain order

Reply:
> I don't know of any code I've seen that needs both order and intersect.

I can come up with some examples of where that might be helpful, but I don't know if I've actually *seen* it in practice, either.  So, yeah, probably not necessary.

> The other question is *which* order is correct. Both lists could have completely different orders so you have to arbitrarily pick on to maintain order.

This isn't a strong enough argument to block anything (particularly given Wake's propensity for pipelines -- it would be the order of the second list), but since there's not a strong argument for maintaining order, I do agree it's moot.